File: SNSPD_TDMS_convert_charGPT.py
"""
Python port of SNSPD_TDMS_convert_charGPT.m
Logic mirrors the Matlab script exactly.
"""

import os
import warnings
import numpy as np
from scipy.io import savemat
from nptdms import TdmsFile

# === 設定區（對應 Matlab 開頭的路徑與參數）===
TDMS_Adress = r'E:\Peaksea\paper\broadbandSMSPD\BroadbandSMSPD_20240519\Laser\P1\20240520\4.5\Pulse\450\10000kHz\5000nW\0degrees\20240520_153645'
Save_Adress = r'E:\Peaksea\paper\broadbandSMSPD\BroadbandSMSPD_20240519\Laser\P1\20240520\4.5\Pulse\450\10000kHz\5000nW\0degrees\20240520_153645'
Exp_para    = 'BroadbandSMSPD_20240519_P1_Pulse_450_5000nW_0degrees_'

# 電壓範圍：對應 Matlab 的 Vol = [500:500:6000, 6100:100:8500]
Vol = np.concatenate([np.arange(500, 6001, 500), np.arange(6100, 8501, 100)])
num_files = len(Vol)

# 預分配 dT
dT = np.zeros(num_files)

# 切換到儲存目錄（對應 cd(Save_Adress)）
os.chdir(Save_Adress)

for ii, voltage in enumerate(Vol):
    filename_TDMS = os.path.join(TDMS_Adress, f"{Exp_para}{voltage}mV.tdms")
    filename_new  = os.path.join(Save_Adress, f"{Exp_para}{voltage}_mV.txt")

    # 檢查檔案是否存在（對應 exist(filename_TDMS, 'file') == 2）
    if not os.path.isfile(filename_TDMS):
        warnings.warn(f"檔案 {filename_TDMS} 不存在，跳過該檔案。")
        continue

    # 讀取 TDMS（對應 convertTDMS）
    tdms = TdmsFile.read(filename_TDMS)

    # 提取信號與觸發（對應 MeasuredData(3) = signal, MeasuredData(4) = trigger）
    group   = tdms['ADC Readout Channels']
    signal  = group['chSig'][:]
    trigger = group['chTrig'][:]

    # 存成 ASCII txt（對應 save(filename_new, 'F', '-ascii')）
    # Matlab 預設 -ascii 格式為 %14.7e (8 位有效數字，科學記號，空白分隔)
    F = np.column_stack([signal, trigger])
    np.savetxt(filename_new, F, fmt='%14.7e', delimiter='   ')

    # 計算 dT（對應 Property(15) / Property(9) * 1e-9）
    # Property(15) = 'record length', Property(9) = 'actual sample rate'
    record_length = float(tdms.properties['record length'])
    sample_rate   = float(tdms.properties['actual sample rate'])
    dT[ii] = record_length / sample_rate * 1e-9  # 奈秒

# 儲存 dT 結果（對應 save('dT_results.mat', 'dT')）
savemat(os.path.join(Save_Adress, 'dT_results.mat'), {'dT': dT})

print('Done')

File: SNSPD_TDMS_convert_charGPT.m
% TDMS轉換toolbox地址 不動
TCT_Adress = 'C:\Users\Peaksea\OneDrive\桌面\台大實驗室\Matlab analysis\github_repo (2)';

% TDMS檔地址
TDMS_Adress = 'E:\Peaksea\paper\broadbandSMSPD\BroadbandSMSPD_20240519\Laser\P1\20240520\4.5\Pulse\450\10000kHz\5000nW\0degrees\20240520_153645';
% 存檔地址
Save_Adress = 'E:\Peaksea\paper\broadbandSMSPD\BroadbandSMSPD_20240519\Laser\P1\20240520\4.5\Pulse\450\10000kHz\5000nW\0degrees\20240520_153645';

% 實驗參數
Exp_para = 'BroadbandSMSPD_20240519_P1_Pulse_450_5000nW_0degrees_';

% 改變當前工作目錄
cd(Save_Adress);
% 添加TDMS轉換工具的路徑
addpath(TCT_Adress);

% 定義電壓範圍
Vol = [500:500:6000, 6100:100:8500];
num_files = length(Vol);

% 預分配記憶體空間
dT = zeros(num_files, 1);

% 迴圈處理每個電壓值
for ii = 1:num_files
    voltage = Vol(ii);
    filename_TDMS = fullfile(TDMS_Adress, [Exp_para, num2str(voltage), 'mV.tdms']);
    filename_new = fullfile(Save_Adress, [Exp_para, num2str(voltage), '_mV.txt']);
    
    % 檢查檔案是否存在
    if exist(filename_TDMS, 'file') == 2
        % 轉換TDMS檔案
        A = convertTDMS(0, filename_TDMS);
        
        % 提取信號和觸發數據
        signal = A.Data.MeasuredData(3).Data;  
        trigger = A.Data.MeasuredData(4).Data;  
        
        % 保存數據到txt文件
        F = [signal, trigger];
        save(filename_new, 'F', '-ascii');
        
        % 計算dT並保存
        dT(ii) = str2double(A.Data.Root.Property(15).Value) / str2double(A.Data.Root.Property(9).Value) * 1e-9; % 奈秒  
        
        % 刪除臨時mat文件
        delete(fullfile(Save_Adress, [Exp_para, num2str(voltage), 'mV.mat']));
    else
        warning('檔案 %s 不存在，跳過該檔案。', filename_TDMS);
    end
end

% 儲存 dT 結果到文件
save(fullfile(Save_Adress, 'dT_results.mat'), 'dT');
